File: snake.cpp
#include <iostream>
#include <cstdlib> // For srand() and rand() functions
#include <ctime>   // For time() function
#include <unistd.h> // For usleep() function

using namespace std;

// Constants
const int WIDTH = 20;
const int HEIGHT = 20;
const char EMPTY = ' ';
const char SNAKE_BODY = 'O';
const char SNAKE_HEAD = '@';
const char FOOD = '*';

// Directions
enum Direction { UP, DOWN, LEFT, RIGHT };

// Game state
bool gameOver;
int score;
Direction dir;

// Snake coordinates
int snakeX[100], snakeY[100];
int snakeLength;

// Food coordinates
int foodX, foodY;

// Setup initial game state
void Setup() {
    gameOver = false;
    score = 0;
    dir = RIGHT;
    snakeX[0] = WIDTH / 2;
    snakeY[0] = HEIGHT / 2;
    snakeLength = 1;
    srand(time(0)); // Seed for random number generator
    foodX = rand() % WIDTH;
    foodY = rand() % HEIGHT;
}

// Draw game board
void Draw() {
    system("clear"); // Clear the console

    // Draw top border
    for (int i = 0; i < WIDTH + 2; i++) {
        cout << "#";
    }
    cout << endl;

    // Draw the board
    for (int y = 0; y < HEIGHT; y++) {
        for (int x = 0; x < WIDTH; x++) {
            if (x == 0)
                cout << "#"; // Draw left border

            if (x == foodX && y == foodY)
                cout << FOOD; // Draw food
            else if (x == snakeX[0] && y == snakeY[0])
                cout << SNAKE_HEAD; // Draw snake head
            else {
                bool print = false;
                for (int k = 1; k < snakeLength; k++) {
                    if (snakeX[k] == x && snakeY[k] == y) {
                        cout << SNAKE_BODY; // Draw snake body
                        print = true;
                    }
                }
                if (!print)
                    cout << EMPTY; // Draw empty space
            }

            if (x == WIDTH - 1)
                cout << "#"; // Draw right border
        }
        cout << endl;
    }

    // Draw bottom border
    for (int i = 0; i < WIDTH + 2; i++) {
        cout << "#";
    }
    cout << endl;

    // Display score
    cout << "Score: " << score << endl;
}

// Input handling
void Input() {
    char input;
    cin >> input;
    switch (input) {
        case 'w':
            dir = UP;
            break;
        case 's':
            dir = DOWN;
            break;
        case 'a':
            dir = LEFT;
            break;
        case 'd':
            dir = RIGHT;
            break;
        case 'x':
            gameOver = true;
            break;
    }
}

// Game logic
void Logic() {
    // Move the snake
    int prevX = snakeX[0];
    int prevY = snakeY[0];
    int prev2X, prev2Y;
    snakeX[0] = (dir == RIGHT) ? snakeX[0] + 1 : (dir == LEFT) ? snakeX[0] - 1 : snakeX[0];
    snakeY[0] = (dir == DOWN) ? snakeY[0] + 1 : (dir == UP) ? snakeY[0] - 1 : snakeY[0];

    // Check collision with borders
    if (snakeX[0] >= WIDTH || snakeX[0] < 0 || snakeY[0] >= HEIGHT || snakeY[0] < 0)
        gameOver = true;

    // Check collision with itself
    for (int i = 1; i < snakeLength; i++) {
        if (snakeX[0] == snakeX[i] && snakeY[0] == snakeY[i])
            gameOver = true;
    }

    // Eat food
    if (snakeX[0] == foodX && snakeY[0] == foodY) {
        score += 10;
        snakeLength++;
        foodX = rand() % WIDTH;
        foodY = rand() % HEIGHT;
    }

    // Move the rest of the snake
    for (int i = 1; i < snakeLength; i++) {
        prev2X = snakeX[i];
        prev2Y = snakeY[i];
        snakeX[i] = prevX;
        snakeY[i] = prevY;
        prevX = prev2X;
        prevY = prev2Y;
    }
}

int main() {
    Setup();

    while (!gameOver) {
        Draw();
        Input();
        Logic();
        usleep(100000); // Delay (100 milliseconds)
    }

    cout << "Game Over! Your final score is: " << score << endl;

    return 0;
}

File: snake.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.util.*;

public class SnakeGame extends JPanel implements KeyListener, ActionListener {
    // Constants
    private static final int WIDTH = 20;
    private static final int HEIGHT = 20;
    private static final int GRID_SIZE = 20;
    private static final int DELAY = 100; // Delay in milliseconds

    // Game state
    private boolean gameOver = false;
    private int score = 0;

    // Snake
    private ArrayList<Point> snake = new ArrayList<Point>();
    private int snakeLength = 3;
    private Point direction = new Point(1, 0); // Initial direction: right

    // Food
    private Point food = new Point();

    public SnakeGame() {
        setPreferredSize(new Dimension(WIDTH * GRID_SIZE, HEIGHT * GRID_SIZE));
        setBackground(Color.BLACK);
        setFocusable(true);
        addKeyListener(this);

        // Initialize snake position
        snake.add(new Point(10, 10));
        snake.add(new Point(9, 10));
        snake.add(new Point(8, 10));

        // Generate initial food position
        generateFood();

        // Start game loop
        Timer timer = new Timer(DELAY, this);
        timer.start();
    }

    // Generate random food position
    private void generateFood() {
        int x = (int) (Math.random() * WIDTH);
        int y = (int) (Math.random() * HEIGHT);
        food.setLocation(x, y);
    }

    // Draw game components
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);

        // Draw snake
        g.setColor(Color.GREEN);
        for (Point p : snake) {
            g.fillRect(p.x * GRID_SIZE, p.y * GRID_SIZE, GRID_SIZE, GRID_SIZE);
        }

        // Draw food
        g.setColor(Color.RED);
        g.fillRect(food.x * GRID_SIZE, food.y * GRID_SIZE, GRID_SIZE, GRID_SIZE);

        // Draw score
        g.setColor(Color.WHITE);
        g.drawString("Score: " + score, 10, 20);

        // Draw game over message
        if (gameOver) {
            g.setColor(Color.WHITE);
            g.setFont(new Font("Arial", Font.BOLD, 24));
            String gameOverMsg = "Game Over! Final Score: " + score;
            int msgWidth = g.getFontMetrics().stringWidth(gameOverMsg);
            g.drawString(gameOverMsg, (WIDTH * GRID_SIZE - msgWidth) / 2, HEIGHT * GRID_SIZE / 
2);
        }
    }

    // Handle keyboard input
    public void keyPressed(KeyEvent e) {
        int key = e.getKeyCode();
        if (key == KeyEvent.VK_UP && direction.y != 1) {
            direction.setLocation(0, -1);
        } else if (key == KeyEvent.VK_DOWN && direction.y != -1) {
            direction.setLocation(0, 1);
        } else if (key == KeyEvent.VK_LEFT && direction.x != 1) {
            direction.setLocation(-1, 0);
        } else if (key == KeyEvent.VK_RIGHT && direction.x != -1) {
            direction.setLocation(1, 0);
        }
    }

    // Unused KeyListener methods
    public void keyReleased(KeyEvent e) {}
    public void keyTyped(KeyEvent e) {}

    // Game loop
    public void actionPerformed(ActionEvent e) {
        if (!gameOver) {
            // Move snake
            Point head = snake.get(0);
            Point newHead = new Point(head.x + direction.x, head.y + direction.y);
            snake.add(0, newHead);

            // Check collision with food
            if (newHead.equals(food)) {
                generateFood();
                score += 10;
            } else {
                snake.remove(snake.size() - 1);
            }

            // Check collision with self or walls
            if (newHead.x < 0 || newHead.x >= WIDTH || newHead.y < 0 || newHead.y >= HEIGHT || 
snake.indexOf(newHead) != 0) {
                gameOver = true;
            }

            // Repaint
            repaint();
        }
    }

    public static void main(String[] args) {
        JFrame frame = new JFrame("Snake Game");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.getContentPane().add(new SnakeGame());
        frame.pack();
        frame.setLocationRelativeTo(null); // Center the frame on the screen
        frame.setVisible(true);
    }
}

